pkg/statistics/table.go

```diff
@@ -662,6 +663,75 @@ func (t *Table) ShallowCopy() *Table {
 	return nt
 }
 
+// CopyForColumnMapUpdate creates a copy optimized for column updates.
+func (t *Table) CopyForColumnMapUpdate() *Table {
+	newHistColl := HistColl{
+		PhysicalID:    t.PhysicalID,
+		RealtimeCount: t.RealtimeCount,
+		columns:       maps.Clone(t.columns),
+		indices:       t.indices,
+		Pseudo:        t.Pseudo,
+		ModifyCount:   t.ModifyCount,
+		StatsVer:      t.StatsVer,
+	}
+	nt := &Table{
+		HistColl:              newHistColl,
+		Version:               t.Version,
+		TblInfoUpdateTS:       t.TblInfoUpdateTS,
+		ExtendedStats:         t.ExtendedStats,
+		ColAndIdxExistenceMap: t.ColAndIdxExistenceMap.Clone(),
+		LastAnalyzeVersion:    t.LastAnalyzeVersion,
+		LastStatsHistVersion:  t.LastStatsHistVersion,
+	}
+	return nt
+}
+
+// CopyForIndexMapUpdate creates a copy optimized for index updates.
+func (t *Table) CopyForIndexMapUpdate() *Table {
+	newHistColl := HistColl{
+		PhysicalID:    t.PhysicalID,
+		RealtimeCount: t.RealtimeCount,
+		columns:       t.columns,
+		indices:       maps.Clone(t.indices),
+		Pseudo:        t.Pseudo,
+		ModifyCount:   t.ModifyCount,
+		StatsVer:      t.StatsVer,
+	}
+	nt := &Table{
+		HistColl:              newHistColl,
+		Version:               t.Version,
+		TblInfoUpdateTS:       t.TblInfoUpdateTS,
+		ExtendedStats:         t.ExtendedStats,
+		ColAndIdxExistenceMap: t.ColAndIdxExistenceMap.Clone(),
+		LastAnalyzeVersion:    t.LastAnalyzeVersion,
+		LastStatsHistVersion:  t.LastStatsHistVersion,
+	}
+	return nt
+}
+
+// CopyForMapsUpdate creates a copy for both col and index map updates.
+func (t *Table) CopyForMapsUpdate() *Table {
```

Comment:
What is the difference between this function and copy? I think they are the same, right?

Reply:
Oh, I see. This actually is a shallow clone for the key and value.

Reply:
I think the 4 similar functions (3 new functions + `ShallowCopy()`) can be simplified to one function: `Table.ShallowCopy(forColumnMapUpdate, forIndexMapUpdate bool)`, and the behavior is that:
- `forColumnMapUpdate` => `maps.Clone(t.columns)`
- `forIndexMapUpdate` => `maps.Clone(t.indices)`
- `forColumnMapUpdate || forIndexMapUpdate` => `t.ColAndIdxExistenceMap.Clone()`
- `!forColumnMapUpdate && !forIndexMapUpdate` => current `ShallowCopy()`

This makes `ShallowCopy()` a bit more complicated, but overall they are easier to understand.

What do you think of it? @Tristan1900 @0xPoe

Reply:
yeah that's a great idea. Will do

Reply:
Good idea. I think we can further merge the Copy and ShallowCopy together. Then, every time somebody wants to use this API, they need to consider which update policy applies to this caller. This would make it difficult to misuse the Copy and ShallowCopy. Every caller has to know what the requirements are here.